src/index.ts

```diff
@@ -160,10 +172,16 @@ export const go = async <T, E extends Error>(
     const attempts = retries ? retries + 1 : 1;
     let lastFailedAttemptResult: GoResultError<E> | null = null;
     for (let i = 0; i < attempts; i++) {
-      // This is guaranteed to be false for the first attempt
-      if (fullTimeoutExceeded) break;
-
+      // Return early in case the global timeout has been exceeded during after attempt wait time.
+      //
+      // This is guaranteed to be false for the first attempt.
+      if (fullTimeoutExceededGoResult) break;
       const goRes = await attempt<T, E>(fn, attemptTimeoutMs);
+      // Return early if the timeout is exceeded not to cause any side effects (such as calling "afterAttemp" function)
+      if (fullTimeoutExceededGoResult) break;
+
+      // Ignore the result of afterFailedAttempt callback
+      if (!goRes.success && afterFailedAttempt) goSync(() => afterFailedAttempt(goRes));
```

Comment:
I would also make sure that the last attempt (successful or not) is not passed to `afterFailedAttempt`. Because I think the common pattern will be something like:
```ts
const logFailedAttempts = (goResult) => console.log(`Attempt error`: ${goResult.error}`);
const goResult = go(() => {...}, {retries: 5, afterFailedAttempt: logFailedAttempts});
if (!goResult.success) console.log(`Final error: ${goResult.error}`);
```
If the final attempt even though failed will be passed to `afterFailedAttempt` it will be logged twice.

Reply:
You could just return there in this case instead of logging (or do a message "all retries to do xyz have failed"). I also find this more "natural" (e.g. by reading just this code I'd assume that the last attempt is logged twice, if we changed it we always need to explicitely log the final attempt).

Reply:
I know but the final attempt is most likely logged anyway. I can't imagine a situation where you would like to log the previous attempts but not the final one.